regression/scripts/optimization_packages/procedure_opt_pack.py

```diff
@@ -0,0 +1,69 @@
+# Procedure.py
+# 
+# Created:  Sep. 2019, M. Clarke
+# Modified: 
+
+# ----------------------------------------------------------------------        
+#   Imports
+# ----------------------------------------------------------------------    
+
+import SUAVE
+from SUAVE.Core import Units, Data
+import numpy as np
+import copy
+from SUAVE.Analyses.Process import Process
+
+
+# ----------------------------------------------------------------------        
+#   Setup
+# ----------------------------------------------------------------------   
+
+def setup():
+    
+    # ------------------------------------------------------------------
+    #   Analysis Procedure
+    # ------------------------------------------------------------------ 
+    
+    # size the base config
+    procedure = Process()
+
+    # post process the results
+    procedure.post_process = post_process
+        
+    # done!
+    return procedure
+
+# ----------------------------------------------------------------------        
+#   Design Mission
+# ----------------------------------------------------------------------    
+def design_mission(nexus):
+    
+    mission = nexus.missions.base
+    results = nexus.results
+    results.base = mission.evaluate()
+    
+    return nexus
+
+# ----------------------------------------------------------------------        
+#   Analysis Setting
+# ----------------------------------------------------------------------   
+
+def post_process(nexus):
+    
+    x1 = nexus.vehicle_configurations.base.x1
+    x2 = nexus.vehicle_configurations.base.x2
+    
+    obj = np.array([x2**2+x1**2]) 
+    
+    # for differential evolution, a penalty function is needed if constraint is violated
```

Comment:
Should this be built into the optimization helper functions rather than the procedure?

Reply:
No, because these penalty functions are user-defined

Reply:
If constraints are meant to be possible with the chosen optimizer, they should be automatically applied based on nexus inputs. This doesn't mean the optimizer can't accept a user defined penalty function, but if this is the case it should be passed into the optimizer in some way. In any case the user shouldn't need to hard code constraint values in the procedure.

Reply:
Differential Evolution (the genetic algorithm) is an unconstrained optimization. I added penalty functions so that I can use the same objective function as particle swarm.

Reply:
Let's discuss the best way to handle this at our next meeting.